src/framework/theme/components/select/select.component.ts

```diff
@@ -0,0 +1,531 @@
+
+  /**
+   * Adds `outline` styles
+   */
+  @Input() outline: boolean;
+
+  /**
+   * Renders select placeholder if nothing selected.
+   * */
+  @Input() placeholder: string = '';
+
+  /**
+   * Will be emitted when selected value changes.
+   * */
+  @Output() selectedChange: EventEmitter<T | T[]> = new EventEmitter();
+
+  /**
+   * Accepts selected item or array of selected items.
+   * */
+  @Input('selected')
+  set _selected(value: T | T[]) {
+    this.writeValue(value);
+  }
+
+  /**
+   * Gives capability just write `multiple` over the element.
+   * */
+  @Input('multiple')
+  set _multiple(multiple: boolean) {
+    this.multiple = convertToBoolProperty(multiple);
+  }
+
+  /**
+   * List of `NbOptionComponent`'s components passed as content.
+   * TODO maybe it would be better provide wrapper
+   * */
+  @ContentChildren(NbOptionComponent, { descendants: true }) options: QueryList<NbOptionComponent<T>>;
+
+  /**
+   * Custom select label, will be rendered instead of default enumeration with coma.
+   * */
+  @ContentChild(NbSelectLabelComponent) customLabel;
+
+  /**
+   * NbCard with options content.
+   * */
+  @ViewChild(NbPortalDirective) portal: NbPortalDirective;
+
+  multiple: boolean = false;
+
+  /**
+   * List of selected options.
+   * */
+  selectionModel: NbOptionComponent<T>[] = [];
+
+  positionStrategy: NbAdjustableConnectedPositionStrategy;
+
+  /**
+   * Current overlay position because of we have to toggle overlayPosition
+   * in [ngClass] direction and this directive can use only string.
+   */
+  overlayPosition: NbPosition = '' as NbPosition;
+
+  /**
+   * Stream of events that will fire when one of the options fire selectionChange event.
+   * */
+  selectionChange: Observable<NbOptionComponent<T>> = defer(() => {
+    return merge(...this.options.map(it => it.selectionChange));
+  });
+
+  protected ref: NbOverlayRef;
+
+  protected alive: boolean = true;
+
+  /**
+   * If a user assigns value before content nb-options's rendered the value will be putted in this variable.
+   * And then applied after content rendered.
+   * Only the last value will be applied.
+   * */
+  protected queue: T | T[];
+
+  /**
+   * Function passed through control value accessor to propagate changes.
+   * */
+  protected onChange: Function = () => {};
+
+  constructor(@Inject(NB_DOCUMENT) protected document,
+              protected overlay: NbOverlayService,
+              protected hostRef: ElementRef,
+              protected positionBuilder: NbPositionBuilderService,
+              protected cd: ChangeDetectorRef) {
+  }
+
+  /**
+   * Determines is select opened.
+   * */
+  get isOpened(): boolean {
+    return this.ref && this.ref.hasAttached();
+  }
+
+  /**
+   * Determines is select hidden.
+   * */
+  get isHidden(): boolean {
+    return !this.isOpened;
+  }
+
+  /**
+   * Returns width of the select button.
+   * */
+  get hostWidth(): number {
+    return this.hostRef.nativeElement.getBoundingClientRect().width;
```

Comment:
is it safe `.getBoundingClientRect()` to call?

Reply:
Yep, because hostRef is injected in the constructor and will be definitely defined.